bundles/org.openhab.binding.bluetooth.listener/src/main/java/org/openhab/binding/bluetooth/listener/internal/ListenerHandler.java

```diff
@@ -0,0 +1,336 @@
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.bluetooth.BeaconBluetoothHandler;
+import org.openhab.binding.bluetooth.BluetoothBindingConstants;
+import org.openhab.binding.bluetooth.BluetoothCharacteristic;
+import org.openhab.binding.bluetooth.notification.BluetoothScanNotification;
+import org.openhab.core.config.core.Configuration;
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.builder.ChannelBuilder;
+import org.openhab.core.thing.binding.builder.ThingBuilder;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.openhab.core.util.HexUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ListenerHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author VitaTucek - Initial contribution
+ */
+@NonNullByDefault
+public class ListenerHandler extends BeaconBluetoothHandler {
+    private final Logger logger = LoggerFactory.getLogger(ListenerHandler.class);
+    private @Nullable ListenerConfiguration config;
+    private final AtomicBoolean receivedStatus = new AtomicBoolean();
+    private @Nullable ScheduledFuture<?> heartbeatFuture;
+    private long scanTime = 0;
+
+    public ListenerHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+        config = getConfigAs(ListenerConfiguration.class);
+        // check configuration
+        int timeout = config != null ? config.dataTimeout : 1;
+        // set unknown
+        updateStatus(ThingStatus.UNKNOWN);
+        // start heartbeat job
+        heartbeatFuture = scheduler.scheduleWithFixedDelay(this::heartbeat, 0, timeout, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Check device connection timeout
+     */
+    private void heartbeat() {
+        if (!receivedStatus.getAndSet(false) && getThing().getStatus() == ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    "@text/offline.communication-error [\"" + (config != null ? config.dataTimeout : "null") + "\"]");
+            scanTime = 0;
+        }
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> heartbeatFuture = this.heartbeatFuture;
+        if (heartbeatFuture != null) {
+            heartbeatFuture.cancel(true);
+            this.heartbeatFuture = null;
+        }
+        super.dispose();
+    }
+
+    @Override
+    public void onScanRecordReceived(BluetoothScanNotification scanNotification) {
```

Comment:
The `onScanRecordReceived` method is long and hard to read/maintain. Maybe some code de-duplication and restructuring can improve the readability. 
```
        processRssiData(scanNotification);
        processManufacturerData(scanNotification);
        processServiceData(scanNotification);
```

and in detail some checks or logic that is used by one or more of those can be extracted to a seperate private helpermethod.